src/main/scala/mesosphere/marathon/upgrade/DeploymentPlanReverter.scala

```diff
@@ -0,0 +1,165 @@
+package mesosphere.marathon.upgrade
+
+import mesosphere.marathon.state.{ Timestamp, AppDefinition, PathId, Group }
+import org.slf4j.LoggerFactory
+
+import scala.collection.immutable.Seq
+
+/**
+  * Contains the logic to revert deployments by calculating the changes
+  * of the deployment and applying the reverse on the given group.
+  */
+private[upgrade] object DeploymentPlanReverter {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  /**
+    * Reverts this plan by applying the reverse changes to the given Group.
+    *
+    * If they were no concurrent deployments and the passed group is equal to
+    * the target group, this method should result in the original group
+    * except for version/timestamp changes.
+    *
+    * If there were concurrent changes, this method tries to revert the changes
+    * of this deployment only without affecting the other deployments.
+    */
+  def revert(original: Group, target: Group): Group => Group = {
+
+    def changesOnIds[T](originalSet: Set[T], targetSet: Set[T])(id: T => PathId): Seq[(Option[T], Option[T])] = {
+      def mapById(entities: Set[T]): Map[PathId, T] =
+        entities.map { entity => id(entity) -> entity }.toMap
+
+      val originalById = mapById(originalSet)
+      val targetById = mapById(targetSet)
+
+      val ids = originalById.keys ++ targetById.keys
+
+      ids
+        .iterator
+        .map { id => originalById.get(id) -> targetById.get(id) }
+        .to[Seq]
+    }
+
+    /* a sequence of tuples with the old and the new group definition (also for unchanged groups) */
+    val groupChanges: Seq[(Option[Group], Option[Group])] =
+      changesOnIds(original.transitiveGroups, target.transitiveGroups)(_.id)
+
+    /* a sequence of tuples with the old and the new app definition */
+    val appChanges: Seq[(Option[AppDefinition], Option[AppDefinition])] = {
+      changesOnIds(original.transitiveApps, target.transitiveApps)(_.id)
+        .filter { case (oldOpt, newOpt) => oldOpt != newOpt }
+    }
+
+    (revertGroupChanges(groupChanges) _).andThen(revertAppChanges(appChanges))
```

Comment:
Apply AppChanges before GroupChanges.
Please make a short comment, that the order is important.